OCPBUGS-32158: Add CEL validation for RegistrySources in Image API

muraee opened the pull request:
HyperShift embeds Image config API directly, so any webhook validation will not be triggered.
This adds CEL validation to the CRD directly to mitigate that.

ref: https://issues.redhat.com/browse/OCPBUGS-32158

muraee commented:
/jira refresh

muraee commented:
/retest-required

muraee commented:
/retest-required

muraee commented:
/retest-required

csrwng commented:
/lgtm

muraee commented:
cc @deads2k

flavianmissi commented:
looks good to me from a registry perspective 👍🏼

JoelSpeed commented:
Can you please link to where this is validated in a webhook for core OCP?

When was this CRD introduced into HyperShift, could there be existing resources that become broken by the addition of this CEL?

muraee commented:
@JoelSpeed not sure where/if the webhook exist, I was just saying, if its exist then it won't run for Hypershift.
I see this validated in the MCO code directly, see:
https://github.com/openshift/machine-config-operator/blob/eeea7495bc40d0f73dd9c2ba030e678ec598d8b1/pkg/controller/container-runtime-config/helpers.go#L504-L506

and its also already mentioned on the field description
https://github.com/openshift/api/blob/85dc560939ef05322ce0b204174e9a02c5376c19/config/v1/types_image.go#L178

muraee commented:
/retest-required

flavianmissi commented:
It doesn't seem to be a webhook... I found two places where this validation is enforced:

- https://github.com/openshift/machine-config-operator/blob/eeea7495bc40d0f73dd9c2ba030e678ec598d8b1/pkg/controller/container-runtime-config/helpers.go#L504-L506
- https://github.com/openshift/openshift-controller-manager/blob/master/pkg/build/controller/build/build_controller.go#L2226-L2228

Although there might be more that I missed 🤔

JoelSpeed commented:
> I don't think they can exist, as this is already not allowed by MCO

How does MCO prevent it? If it's not via an admission time check (webhook?) then the resources absolutely can exist. They may break MCO but they could exist in the etcd DB. Ratcheting validation may save you, but you'll need the management cluster to be on 4.18 for that, which I expect won't be happening any time soon?

JoelSpeed commented:
> It doesn't seem to be a webhook... I found two places where this validation is enforced:

> https://github.com/openshift/machine-config-operator/blob/eeea7495bc40d0f73dd9c2ba030e678ec598d8b1/pkg/controller/container-runtime-config/helpers.go#L504-L506
> https://github.com/openshift/openshift-controller-manager/blob/master/pkg/build/controller/build/build_controller.go#L2226-L2228
> Although there might be more that I missed 🤔

Would any of this cause the MCO to do degraded and become not-upgradeable?

JoelSpeed commented on config/v1/types_image.go:
```diff
@@ -162,6 +162,8 @@ type RegistryLocation struct {
 
 // RegistrySources holds cluster-wide information about how to handle the registries config.
+//
+// +kubebuilder:validation:XValidation:rule="self.blockedRegistries.size() > 0 ? self.allowedRegistries.size() == 0 : true",message="Only one of BlockedRegistries or AllowedRegistries may be set"
```
Would this be an easier validation to understand? 
```suggestion
// +kubebuilder:validation:XValidation:rule="self.blockedRegistries.size() == 0 || self.allowedRegistries.size() == 0",message="Only one of BlockedRegistries or AllowedRegistries may be set"
```

muraee replied:
updated

muraee commented:
/retest-required

flavianmissi commented:
> Would any of this cause the MCO to do degraded and become not-upgradeable?

I don't know, but I asked here: https://redhat-internal.slack.com/archives/C02CZNQHGN8/p1727348206792439

yuqi-zhang commented:
Tested briefly on a 4.18 cluster. It looks like it won't degrade anything in the MCO directly, but I'm pretty sure it will degrade when you upgrade the cluster (i.e. when a new controller rolls out.)

What happens is now the controller syncs will continuously error out on the broken image.config 
```Error syncing image config openshift-config: could not Create/Update MachineConfig: could not update policy json with new changes: invalid images config: only one of AllowedRegistries or BlockedRegistries may be specified``` but nothing is syncing that to the pool/operator objects.

When the MCO upgrades, it will roll out a new controller, and the MCO will expect all MachineConfigs managed by the controllers (in this case, the registries configuration is what image.config eventually populates) to be updated to the new controller version. The containerruntimeconfigcontroller won't be able to do so for the image.config changes, so the MCO will eventually degrade and complain that there's a controller version mismatch for a machineconfig that it manages.

Interestingly it does look like openshift-samples degrade:
```
openshift-samples                          4.18.0-0.nightly-2024-09-23-182657   True        True          True       30h     Samples installation in error at 4.18.0-0.nightly-2024-09-23-182657: &errors.errorString{s:"global openshift image configuration prevents the creation of imagestreams using the registry "}
```

JoelSpeed commented:
> but I'm pretty sure it will degrade when you upgrade the cluster

Did you verify this? It sounds to me like this should be a pre-upgrade check and that the operator should report upgradeable false if it determines this scenario, or, can it not be determined prior to the upgrade being initiated because all of the existing MCs are at the correct version before this change was initiated?

yuqi-zhang commented:
@djoshy and I tested this, and indeed if the controller version changes (i.e. an upgrade), the pools become degraded and upgradeable=false gets set. Interestingly the CO itself doesn't seem to degrade, which I've opened a side bug here: https://issues.redhat.com/browse/OCPBUGS-42546

> It sounds to me like this should be a pre-upgrade check and that the operator should report upgradeable false if it determines this scenario, or, can it not be determined prior to the upgrade being initiated because all of the existing MCs are at the correct version before this change was initiated?

The current setup is that until a new controller rolls out, the render controller isn't able to determine this. That said, the containerruntimeconfigcontroller should either be able to degrade the pool or the CO when the failure initially happens, so we don't get to this stage I think. If we're in consensus there, I can create a card to track adding that failure mode.

JoelSpeed commented:
Ok, I think we should proceed with the CEL, but I'd like to try and test the ratcheting scenario. Ratcheting validation was added in 1.30, so now that OCP is based on 1.31, we should be able to leverage it safely (delayed a release because CRDs are upgraded prior to API servers in some cases)

muraee commented:
@JoelSpeed can we process with this PR, or do you want to use the `Ratcheting validation` you mentioned instead?

JoelSpeed commented:
This validation should ratchet already, what I'd like to see is a test case added that shows this. Check the `tests` folder readme on how to set up a ratcheting validation integration test for this. https://github.com/openshift/api/pull/2142 also provides some examples of adding ratcheting tests

muraee commented:
@JoelSpeed I added some tests, and adjusted the CEL to make the fields mutually exclusive. `self.blockedRegistries.size() == 0 || self.allowedRegistries.size() == 0` allowed both fields to be set and made at least one of them required, whereas it's allowed for both to be empty.

muraee commented on config/v1/tests/images.config.openshift.io/AAA_ungated.yaml:
```diff
@@ -12,3 +12,93 @@ tests:
+        kind: Image
+        spec:
+          registrySources:
+            allowedRegistries: ["test"]
```
done

JoelSpeed commented:
/lgtm

muraee commented:
/retest-required

muraee commented:
/retest-required

muraee commented:
/cherry-pick release-4.18
